database2/ffboltdb/db.go

```diff
@@ -0,0 +1,1510 @@
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// NOTE: The data returned by this function is only valid during a database
+// transaction.  Attempting to access it after a transaction has ended results
+// in undefined behavior.  This constraint prevents additional data copies and
+// allows support for memory-mapped database implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlockHeaders(hashes []wire.ShaHash) ([][]byte, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	// NOTE: This could check for the existence of all blocks before loading
+	// any of the headers which would be faster in the failure case, however
+	// callers will not typically be calling this function with invalid
+	// values, so optimize for the common case.
+
+	// Load the headers.
+	headers := make([][]byte, len(hashes))
+	for i := range hashes {
+		hash := &hashes[i]
+
+		// When the block is pending to be written on commit return the
+		// bytes from there.
+		if tx.pendingBlocks != nil {
+			if idx, exists := tx.pendingBlocks[*hash]; exists {
+				blkBytes := tx.pendingBlockData[idx].bytes
+				headers[i] = blkBytes[0:blockHdrSize:blockHdrSize]
+				continue
+			}
+		}
+
+		// Fetch the block index row and slice off the header.  Notice
+		// the use of the cap on the subslice to prevent the caller
+		// from accidentally appending into the db data.
+		blockRow, err := tx.fetchBlockRow(hash)
+		if err != nil {
+			return nil, err
+		}
+		endOffset := blockLocSize + blockHdrSize
+		headers[i] = blockRow[blockLocSize:endOffset:endOffset]
+	}
+
+	return headers, nil
+}
+
+// FetchBlock returns the raw serialized bytes for the block identified by the
+// given hash.  The raw bytes are in the format returned by Serialize on a
+// wire.MsgBlock.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockNotFound if the request block hash does not exist
+//   - ErrTxClosed if the transaction has already been closed
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// In addition, returns ErrDriverSpecific if any failures occur when reading the
+// block files.
+//
+// NOTE: The data returned by this function is only valid during a database
+// transaction.  Attempting to access it after a transaction has ended results
+// in undefined behavior.  This constraint prevents additional data copies and
+// allows support for memory-mapped database implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlock(hash *wire.ShaHash) ([]byte, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	// When the block is pending to be written on commit return the bytes
+	// from there.
+	if tx.pendingBlocks != nil {
```

Comment:
Unneeded check, I'd search the code for all of these and remove all of the checks when the map is not assigned to.